core/src/main/java/org/infinispan/protostream/impl/RawByteArrayOutputStreamImpl.java

```diff
@@ -0,0 +1,93 @@
+package org.infinispan.protostream.impl;
+
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+
+import net.jcip.annotations.NotThreadSafe;
+
+@NotThreadSafe
+public class RawByteArrayOutputStreamImpl extends OutputStream implements org.infinispan.protostream.RawByteArrayOutputStream {
+
+   static final int MIN_SIZE = 32;
+
+   static final int DEFAULT_DOUBLING_SIZE = 4 * 1024 * 1024; // 4MB
+
+   byte[] buf;
+   int pos = 0;
+
+   public RawByteArrayOutputStreamImpl() {
+   }
```

Comment:
shouldn't create a new a byte[] with default size (`MIN_SIZE`)? We could skip the null checks in `ensureCapacity`and `getByteBuffer`.

Reply:
During my benchmarks it seem liked initializing the array was leading to worse performance. I believe this is because we're creating a lot of stream instances and in many cases the initial capacity required exceeded `MIN_SIZE`.